Mathlib/Algebra/Category/ModuleCat/Injective.lean

```diff
@@ -0,0 +1,35 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.Algebra.Category.ModuleCat.Abelian
+import Mathlib.Algebra.Category.ModuleCat.ChangeOfRings
+import Mathlib.Algebra.Category.GroupCat.Abelian
+import Mathlib.Algebra.Category.GroupCat.Injective
+import Mathlib.Algebra.Category.GroupCat.ZModuleEquivalence
+
+/-!
+# Category of $R$-modules has enough injectives
+
+we lift enough injectives of abelian groups to arbitrary $R$-modules by adjoint functors
+`restrictScalars ⊣ coextendScalars`
+
+## Implementation notes
+This file is not part of `Algebra/Module/Injective.lean` to prevent import loop: enough-injectives
+of abelian groups needs `Algebra/Module/Injective.lean` and this file needs enough-injectives of
+abelian groups.
+-/
+
+open CategoryTheory
+
+universe u v
+variable (R : Type u) [Ring R]
+
+instance : EnoughInjectives (ModuleCat.{max v u} R) :=
```

Comment:
Indeed, the following does **not** work
```lean
instance : EnoughInjectives (ModuleCat.{u} R) := inferInstance
```
where `R` lives in universe `u` as well

Reply:
Does this mean `ModuleCat.restrictCoextendScalarsAdj` need to use `UnivLE` also?

Reply:
> Indeed, the following does **not** work
```lean
instance : EnoughInjectives (ModuleCat.{u, u} R) := inferInstance
```
> where `R` lives in universe `u` as well

Yeah, Lean 4 refuses to unify `max u v` with `u` because `v` can be either `0` or `u`. This `{u,u}` version could be an instance (as well as `{u+1, u}`, `{u, 0}`, but they will subsumed by the instance with UnivLE).

> Does this mean `ModuleCat.restrictCoextendScalarsAdj` need to use `UnivLE` also?

Not necessarily, as you can specify the universes explicitly when using it.

Reply:
Can you elaborate please, I tried the following
```lean
instance [UnivLE.{u, v}] : EnoughInjectives (ModuleCat.{v, u} R) := 
  EnoughInjectives.of_adjunction (ModuleCat.restrictCoextendScalarsAdj.{v, 0, u} (algebraMap ℤ R))
```
Looks like that Lean 4 doesn't see `Type v` and `Type {max v u}` are the same.

A minimal example is perhaps this:
```lean
example [UnivLE.{u, v}] : ModuleCat.{v} ℤ ⥤ ModuleCat.{v, u} R :=
  ModuleCat.coextendScalars.{0, u, v} (algebraMap ℤ R)
```
does not compile

Reply:
`UnivLE.{u, v}` says every type in universe u is in bijection with some type in universe v (after my refactor #8556). It doesn't give you `v = max u w` for some `w` (and I don't think it's currently possible to state this in Lean). It seems a bit nontrival to transfer EnoughInjectives from `max u v` to `UnivLE`, so I suggest you just leave it as a `def` for now and manually specify the universes when you want to use it.

Reply:
Thanks for the suggestion. I have changed instance to a def. I realized that we know `Module.Baer` implies injectivity but not the otherway around.

Reply:
#8734